PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
+        """
+        Remove the passed modules and hence all hook scripts cached for these
+        modules from this cache.
+
+        Parameters
+        ----------
+        module_names : list
+            List of all fully-qualified module names to be removed.
+        """
+
+        for module_name in set(module_names):  # Eliminate duplicate entries.
+            if module_name in self:
+                del self[module_name]
+
+
+class ModuleHook(object):
+    """
+    Cached object encapsulating a lazy loadable hook script.
+
+    This object's `code` property provides the currently loaded contents (e.g.,
+    attributes, functions) of this hook script. The first access of this
+    property silently lazily loads this script into an in-memory code object.
+
+    Parameters
+    ----------
+    module_graph : ModuleGraph
+        Current module graph.
+    module_name : str
+        Name of the module hooked by this hook script.
+    hook_filename : str
+        Absolute or relative path of this hook script.
+    hook_module_name : str
+        Name of the in-memory module of this hook script's interpreted contents.
+    _scope : module
+        In-memory module of this hook script's interpreted contents, lazily
+        loaded on the first access of the `scope` property _or_ `None` if this
+        property has yet to be accessed.
+    """
+
+    ## Private methods
+
+    def __init__(
+        self,
+        module_graph,
+        module_name,
+        hook_filename,
+        hook_module_name_prefix,
+    ):
+        """
+        Initialize this metadata.
+
+        Parameters
+        ----------
+        module_graph : ModuleGraph
+            Current module graph.
+        module_name : str
+            Name of the module hooked by this hook script.
+        hook_filename : str
+            Absolute or relative path of this hook script.
+        hook_module_name_prefix : str
+            String prefixing the name of the in-memory module for this hook
+            script's code object. To avoid namespace clashes with similar
+            modules created by other `ModuleHook` objects in other
+            `ModuleHookCache` containers, this string _must_ be unique to the
+            `ModuleHookCache` container containing this `ModuleHook` object. If
+            this string is non-unique, an existing in-memory module will be
+            erroneously reused when lazily loading this hook script, thus
+            erroneously resanitizing previously sanitized hook script attributes
+            (e.g., `datas`) with the `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
+        # avoiding circular reference issues. See ModuleHookCache.__init__().
+        self.module_graph = module_graph
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
```

Comment:
> Please be more verbose on why we need this.

Verbosity is my middle name. <sup>It's actually William.</sup>

We need this because, without this, **tests fail.** This was actually the most difficult aspect of this pull request to debug, and I should have documented why in detail. It took me several days to eventually realize what was happening. _Which really sucked._

So, here's what was happening. Tests would individually succeed but fail when run as a group. For every test `test_poo()`, the corresponding:

* `test_poo()[onedir]` test would succeed both when run individually _and_ when followed by a `test_poo()[onefile]` test.
* `test_poo()[onefile]` test would succeed _only_ when run individually. When preceded by a `test_poo()[onedir]` test, a `test_poo()[onefile]` test would **always fail** with a `FileNotFoundError` raised by the `PyInstaller.building.utils.format_binaries_and_datas()` helper function.

Hence, output resembling:

```
$ py.test -k 'test_poo[onedir]'
tests/functional/test_libraries.py::test_poo[onedir] PASSED
$ py.test -k 'test_poo[onefile]'
tests/functional/test_libraries.py::test_poo[onefile] PASSED
$ py.test -k test_poo
tests/functional/test_libraries.py::test_poo[onedir] PASSED
tests/functional/test_libraries.py::test_poo[onefile] FAILED
```

_wat?_ This is where I began tearing my thinning hair out.

To explain this heinousness, let's first talk about the `format_binaries_and_datas()` function. This function receives as input a list of 2-tuples `(source_dir_or_glob, target_dir)` and returns as output a list of 2-tuples `(target_file, source_file)`. A list returned by this function _cannot_ be passed back to this function, as the elements of the 2-tuples of the former are in the reverse order of the 2-tuples of the latter. Doing so reliably raises the `FileNotFoundError` error raised by the failing test above. In pseudo-Python:

```
>>> output = format_binaries_and_datas([('/usr', 'poo')])
>>> output
[('poo', '/usr')]
>>> format_binaries_and_datas(output)
FileNotFoundError: Path or glob "poo" not found or matches no files.
```

So far, so good. Since the failing test above raises the same `FileNotFoundError`, it stood to reason that the `format_binaries_and_datas()` function was receiving a list that had _already_ been returned by a prior call to that function. But how could that be? That function is only called twice in the codebase: once by **(A)** `Analysis.__init__()` to handle additional `binaries` and `datas` defined by user `.spec` files and once by **(B)** `ModuleHook._sanitize_hook_attrs()` to handle `binaries` and `datas` defined by the current hook script.

So, what gives? The `format_binaries_and_datas()` function is _never_ called twice for the same hook script. That function should _never_ receive a list output by a prior call to that function. **And then I realized with horror.**

Previously, each hook script was _always_ loaded into the same in-memory module `PyInstaller_hook_{modname}` by the `importlib_load_source()` helper function. If this function were (for whatever reason) called twice with the same module name in the same Python process, the second call would attempt to reload the previously loaded hook script back into that module – which already exists. Is that problematic?

**It is.** At least under Python 3, where `importlib_load_source()` reduces to `importlib.machinery.SourceFileLoader(modname, pathname).load_module()`. For whatever reason, that logic behaves erratically when passed the name of a module that's already been loaded. I didn't delve into the specifics, but this logic appears to silently ignore attempts to reload already loaded modules. _Isn't that special._

If `py.test` runs the `test_poo()[onedir]` and `test_poo()[onefile]` tests from the same Python process (as it appears to) rather than forking a new process for each (as I assumed it did), these two tests will share the same set of in-memory modules. Due to the above `importlib_load_source()` behaviour, each `ModuleHook` object loaded for the second `test_poo()[onefile]` test will simply reuse the previously cached hook script module rather than reloading that module.

Is there a point to any of this? **Yes.** _Sort of._ Any `datas` and `binaries` attributes previously set by the first `test_poo()[onedir]` test would then be reused by the second `test_poo()[onefile]` test – except those attributes were already sanitized by `format_binaries_and_datas()`. The second `test_poo()[onefile]` test doesn't know that, of course, so it tries to resanitize them.

And this is why I didn't document this in the code in detail.